<a href="https://colab.research.google.com/github/xCHMNx/Math152/blob/main/P4M_Notebook1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#  Computing with Python:  an Introduction

Welcome to programming!

What is the difference between *Python* and a calculator?  We begin this first lesson by showing how Python can be used **as** a calculator, and we move into one of the most important programming structures -- the **loop**.  Loops allow computers to carry out repetetive computations, with just a few commands.

To navigate this notebook, you might wish to click on the Table of Contents icon (near top-left of the Google Colab window).  Then you can expand the sections to work on each one.

Remember to save your work frequently on your personal GitHub repository.

You do **not** need to complete every exercise to get full credit for this assignment.  Please **do** try to make it to the end of the notebook to learn all the concepts, and attempt as many exercises as time permits.

# Python as a calculator

Different kinds of data are stored as different *types* in Python.  For example, if you wish to work with integers, your data is typically stored as an *int*.  A real number might be stored as a *float*.  There are types for booleans (True/False data), strings (like "Hello World!"), and many more we will see.  

A more complete reference for Python's numerical types and arithmetic operations can be found in the [official Python documentation](https://docs.python.org/3/library/stdtypes.html).  The [official Python tutorial](https://docs.python.org/3/tutorial/introduction.html) is also a great place to start.

Python allows you to perform arithmetic operations:  addition, subtraction, multiplication, and division, on numerical types.  The operation symbols are `+`, `-`, `*`, and `/`.   Evaluate each of the following cells to see how Python performs operations on *integers*.  

To evaluate the cell, click anywhere within the cell to select it and use the keyboard shortcut *Shift-Enter* to evaluate.  But as you go through this and later lessons, try to *predict* what will happen when you evaluate the cell before you hit Shift-Enter.  Don't just click click click!  Think for a few seconds or a minute, and then Shift-Enter.

In [ ]:
2 + 3

5

In [ ]:
2 * 3

6

In [ ]:
5 - 11

-6

In [ ]:
5.0 - 11

-6.0

In [ ]:
5 / 11

0.45454545454545453

In [ ]:
6 / 3

2.0

In [ ]:
5 // 11

0

In [ ]:
6 // 3

2

The results are probably not too surprising, though the last two require a bit of explanation.  Python *interprets* the input number 5 as an *int* (integer) and 5.0 as a *float*.  "Float" stands for "floating point number," which are decimal approximations to real numbers.  The word "float" refers to the fact that the decimal (or binary, for computers) point can float around (as in 1.2345 or 12.345 or 123.45 or 1234.5 or 0.00012345).  There are deep computational issues related to how computers handle decimal approximations, and you can [read about the IEEE standards](https://en.wikipedia.org/wiki/IEEE_754) if you're interested.

Python enables different kinds of division.  The single-slash division in Python 3.x gives a floating point approximation of the quotient.  That's why `5 / 11` and `6 / 3` both output floats.  On the other hand, `5 // 11` and `6 // 3` yield integer outputs (rounding down) -- this is useful, but one has to be careful!

In fact the designers of Python changed their mind.  **This tutorial assumes that you are using Python 3.x.**  If you are using Python 2.x, the command `5 / 11` would output zero.

In [ ]:
-12 // 5  # What will this output?  Guess before evaluating!

-3

Why use integer division `//` and why use floating point division?  In practice, integer division is typically a faster operation.  So if you only need the rounded result (and that will often be the case), use integer division.  It will run much faster than carrying out floating point division then manually rounding down.

Observe that floating point operations involve approximation.  The result of `5.0/11.0` might not be what you expect in the last digit.  Over time, especially with repeated operations, *floating point approximation* errors can add up!

You might be wondering about the little [XX] and [XX] prompts in the code cells above.  What is their purpose?  Guess what the following line will do.

In [ ]:
Out[4] + Out[5]

-5.545454545454546

Cool, huh?  It's nice to have a record of previous computations, especially if you don't want to type something again.

Python allows you to group expressions with parentheses, and follows the order of operations that you learn in school.

In [ ]:
(3 + 4) * 5

35

In [ ]:
3 + (4 * 5)

23

In [ ]:
3 + 4 * 5   #  What do you think will be the result?  Remember PEMDAS?

23

In [ ]:
4(1+3) # Python does not use "implicit" multiplication!  This should give an error.

TypeError: ignored

Now is a good time to try a few computations of your own, in the empty cell below.  You can type any Python commands you want in the empty cell.  If you want to insert a new cell into this notebook, just hover your cursor in the space between cells in Colab.

In [ ]:
#  An empty cell.  Have fun!
print(11**3)
print(1331**(1/3))

1331
10.999999999999998


For number theory, *division with remainder* is an operation of central importance.  Integer division provides the quotient, and the operation `%` provides the remainder.  It's a bit strange that the percent symbol is used for the remainder, but this [dates at least to the early 1970s](https://softwareengineering.stackexchange.com/questions/294297/in-what-programming-language-did-the-use-of-the-percent-sign-to-mean-modulo) and has become standard across computer languages.

In [ ]:
23 // 5  # Integer division

4

In [ ]:
23 % 5  # The remainder after division

3

Note in the code above, there are little "comments".  To place a short comment on a line of code, just put a hashtag `#` at the end of the line of code, followed by your comment.

Python gives a single command for division with remainder.  Its output is a *tuple*.

In [ ]:
divmod(23,5)

(4, 3)

In [ ]:
type(divmod(23,5))

tuple

All data in Python has a type, but a common complaint about Python is that types are a bit concealed "under the hood".  But they are not far under the hood!  Anyone can find out the type of some data with a single command.

In [ ]:
type(3)

int

In [ ]:
type(3.0)

float

In [ ]:
type('Hello')

str

In [ ]:
type([1,2,3])

list

The key to careful computation in Python is always being *aware of the type* of your data, and *knowing* how Python operates differently on data of different types.

In [ ]:
3 + 3

6

In [ ]:
3.0 + 3.0

6.0

In [ ]:
'Hello' + 'World!'

'HelloWorld!'

In [ ]:
[1,2,3] + [4,5,6]

[1, 2, 3, 4, 5, 6]

In [ ]:
3 + 3.0

6.0

In [ ]:
3 + 'Hello!'  # Uh oh!

TypeError: ignored

In [ ]:
#  An empty cell.  Have fun!
#  Try operating on ints, floats, and strings, with different operations.  Which ones work?  How?
print("Hello"*3)
print(5**2)
print(5.0**2)
print([1, 2, 3]*2)

HelloHelloHello
25
25.0
[1, 2, 3, 1, 2, 3]


As you can see, addition (the `+` operator) is interpreted differently in the contexts of numbers, strings, and lists.  The designers of Python allowed us to add *numbers* of different types:  if you try to operate on an *int* and a *float*, the *int* will typically be *coerced* into a float in order to perform the operation.  But the designers of Python did not give meaning to the addition of a number with a string, for example.  That's why you probably received a *TypeError* after trying to add a number to a string. 

On the other hand, Python does interpret *multiplication* of a natural number with a string or a list.

In [ ]:
3 * 'Hello!'

'Hello!Hello!Hello!'

In [ ]:
0 * 'Hello!'

''

In [ ]:
2 * [1,2,3]

[1, 2, 3, 1, 2, 3]

Can you create a string with 100 A's (like `AAA...`)?  Use an appropriate operation in the cell below.

In [ ]:
#  Practice cell
'A'*100

'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA'

Exponents in Python are given by the `**` operator.  The following lines compute 2 to the 1000th power, in two different ways.

In [ ]:
2**1000

10715086071862673209484250490600018105614048117055336074437503883703510511249361224931983788156958581275946729175531468251871452856923140435984577574698574803934567774824230985421074605062371141877954182153046474983581941267398767559165543946077062914571196477686542167660429831652624386837205668069376

In [ ]:
2.0**1000

1.0715086071862673e+301

As before, Python interprets an operation (`**`) differently in different contexts.  When given integer input, Python evaluates `2**1000` **exactly**.  The result is a large integer.  A nice fact about Python, for mathematicians, is that it handles exact integers of arbitrary length!  Many other programming languages (like C++) will give an error message if integers get too large in the midst of a computation.  

New in version 3.x, Python implements long integers without giving signals to the programmer or changing types.  In Python 2.x, there were two types: *int* for somewhat small integers (e.g., up to $2^{31}$) and *long* type for all larger integers.  Python 2.x would signal which type of integer was being used, by placing the letter "L" at the end of a long integer.  Now, in Python 3.x, the programmer doesn't really see the difference.  There is only the *int* type.  But Python still optimizes computations, using hardware functionality for arithmetic of small integers and custom routines for large integers.  The programmer doesn't have to worry about it most of the time.

For scientific applications, one often wants to keep track of only a certain number of significant digits (sig figs).  If one computes the floating point exponent `2.0**1000`, the result is a decimal approximation.  It is still a float.  The expression "e+301" stands for "multiplied by 10 to the 301st power", i.e., Python uses *scientific notation* for large floats.

In [ ]:
type(2**1000)

int

In [ ]:
type(2.0**1000)

float

In [ ]:
#  An empty cell.  Have fun!
len(list(str(2**1000))) # Digits in the integer 2**1000

302

Now is a good time for reflection.  Double-click in the cell below to answer the given questions.  Cells like this one are used for text rather than Python code.  Text is entered using *markdown*, but you can typically just enter text as you would in any text editor without problems.  Press *shift-Enter* after editing a text cell to complete the editing process.  

### Exercises

1.  What data types have you seen, and what kinds of data are they used for?  Can you remember them without looking back?

2.  How is division `/` interpreted differently for different types of data?

3.  How is multiplication `*` interpreted differently for different types of data?

4.  What is the difference between 100 and 100.0, for Python?

Double-click this text cell to edit it, and answer the exercises.  This may be graded, so please complete all questions!  Write in clear, complete, and concise sentences.

1. We have seen integers, floats, strings, lists, and booleans. Integers and floats are used for storing numbers, strings are used to store letters and symbols (and sometimes numbers when you want to write them out), and lists are used to store ordered collections of objects. Booleans store one of two values, either *True* or *False*.

2. Division (with one `/`) is interpreted as a floating point operation between any two numbers and will return a floating point approximation. It cannot be used between strings and lists.

3. Multiplication (with one `*`) is interpreted differently depending on the data types. If both are integers, the output will be an integer. If one or more are floating point numbers, the output will also be a float. This operation can also be done between integers and strings or lists, but not between floats and strings or lists. Between an integer **n** and a string **S**, the output is a string containing **n** copies of the string **S**. Between an integer **n** and a list **L** the same thing happens, but the output is a list containing the elements of **L** repeated **n** times. This operation cannot be used between lists or strings.

4. The number 100 is an integer and the number 100.0 is a float. These will behave a bit differently with some operations, like multiplication described above.

# Calculating with booleans

A *boolean* (type *bool*) is the smallest possible piece of data.  While an *int* can be any integer, positive or negative, a *boolean* can only be one of two things:  *True* or *False*.  In this way, booleans are useful for storing the answers to yes/no questions.  

Questions about (in)equality of numbers are answered in Python by *operations* with numerical input and boolean output.  Here are some examples.  A more complete reference is [in the official Python documentation](https://docs.python.org/3/library/stdtypes.html#boolean-operations-and-or-not).

In [ ]:
3 > 2

True

In [ ]:
type(3 > 2)

bool

In [ ]:
10 < 3

False

In [ ]:
2.4 < 2.4000001

True

In [ ]:
32 >= 32

True

In [ ]:
32 >= 31

True

In [ ]:
2 + 2 == 4

True

Which number is bigger:  $23^{32}$ or $32^{23}$?  Use the cell below to answer the question!

In [ ]:
#  Write your code here.
23**32 > 32**23 # 23**32 is larger!

True

The expressions `<`, `>`, `<=`, `>=` are interpreted here as **operations** with numerical input and boolean output.  The symbol `==` (two equal symbols!) gives a True result if the numbers are equal, and False if the numbers are not equal.  An extremely common typo is to confuse `=` with `==`.  But the single equality symbol `=` has an entirely different meaning, as we shall see.

Using the remainder operator `%` and equality, we obtain a divisibility test.

In [ ]:
63 % 7 == 0  # Is 63 divisible by 7?

True

In [ ]:
101 % 2 == 0  # Is 101 even?

False

Use the cell below to determine whether 1234567890 is divisible by 3.

In [ ]:
# Your code goes here.
1234567890 % 3 == 0 # It is divisible by 3!

True

Booleans can be operated on by the standard logical operations: and, or, not.  In ordinary English usage, "and" and "or" are conjunctions, while here in *Boolean algebra*, "and" and "or" are operations with Boolean inputs and Boolean output.  The precise meanings of "and" and "or" are given by the following **truth tables**.

    
| and | True | False |
|-----|------|-------|
| **True** | True | False |
| **False** | False | False|

<br>
  
| or | True | False |
|-----|------|-------|
| **True** | True | True |
| **False** | True | False|

In [ ]:
True and False

False

In [ ]:
True or False

True

In [ ]:
True or True

True

In [ ]:
not True

False

Use the truth tables to predict the result (True or False) of each of the following, before evaluating the code.

In [ ]:
(2 > 3) and (3 > 2)

False

In [ ]:
(1 + 1 == 2) or (1 + 1 == 3)

True

In [ ]:
not (-1 + 1 >= 0)

False

In [ ]:
2 + 2 == 4

True

In [ ]:
2 + 2 != 4  # For "not equal", Python uses the operation `!=`.

False

In [ ]:
2 + 2 != 5  # Is 2+2 *not* equal to 5?

True

In [ ]:
not (2 + 2 == 5)  # The same as above, but a bit longer to write.

True

Experiment below to see how Python handles a double or triple negative, i.e., something with a `not` `not`.

In [ ]:
# Experiment here.
print(1 < 2)
print(not(1 < 2))
print(not(not(1 < 2)))
print(not(not(not(1 < 2))))

True
False
True
False


Python does give an interpretation to arithmetic operations with booleans and numbers.  Try to guess this interpretation with the following examples.  Change the examples to experiment!

In [ ]:
False * 100

0

In [ ]:
True + 13

14

In [ ]:
# It looks like Python follows the interpretation that False can be represented by 0 and True can be represented by 1
print(True ** 20)
print(False + 10)

1
10


In [ ]:
# Experiment with and, or, not
print(not False or True) # False if or has higher precedence, True if not has higher precedence (returns True)
print(False and True or True) # False if or has higher precedence, True if and has higher precedence (returns True)
#print(True or True and False) # Check if left to right order affects the outcome (it doesn't)
print(True and not False) # In this case, the not operator MUST be evaluated before the and operator, otherwise the expression could not be parsed

# In conclusion, the not operator has the highest priority, then the and operator, then the or operator

True
True
True


In [ ]:
def xor(bool1, bool2):
  return (bool1 or bool2) and not (bool1 and bool2)

In [ ]:
# Test xor
for bool1 in [True, False]:
  for bool2 in [True, False]:
    print(bool1, bool2)
    print(xor(bool1, bool2))
    print()

True True
False

True False
True

False True
True

False False
False



This ability of Python to interpret operations based on context is a mixed blessing.  On one hand, it leads to handy shortcuts -- quick ways of writing complicated programs.  On the other hand, it can lead to code that is harder to read, especially for a Python novice.  Good programmers aim for code that is easy to read, not just short!

The [Zen of Python](https://www.python.org/dev/peps/pep-0020/) is a series of 20 aphorisms for Python programmers.  The first seven are below.

> Beautiful is better than ugly.

> Explicit is better than implicit.

> Simple is better than complex.

> Complex is better than complicated.

> Flat is better than nested.

> Sparse is better than dense.

> Readability counts.

### Exercises

1.  Did you look at the truth tables closely?  Can you remember, from memory, what `True or False` equals, or what `True and False` equals?  

2.  How might you easily remember the truth tables?  How do they resemble the standard English usage of the words "and" and "or"?

3.  If you wanted to know whether a number, like 2349872348723, is a multiple of 7 but **not** a multiple of 11, how might you write this in one line of Python code?

4.  You can chain together `and` commands, e.g., with an expression like `True and True and True` (which would evaluate to `True`).  You can also group booleans, e.g., with `True and (True or False)`.  Experiment to figure out the order of operations (`and`, `or`, `not`) for booleans.

6.  The operation `xor` means "exclusive or".  Its truth table is: `True xor True = False` and `False xor False = False` and `True xor False = True` and `False xor True = True`.  How might you implement `xor` in terms of the usual `and`, `or`, and `not`?



###  Solutions

(Edit here to give solutions to the exercises)

1. The expression `True or False` would equal `True` because only one argument has to be `True` for a `True` result with `or`. The expression `True and False` would equal `False` because both arguments have to be `True` for a `True` result with `and`.

2. In English, the word "and" typically implies both things on either side are important. The word "or" typically implies one thing or the other is important. This closely matches the mathematical usage of "and" and "or" mentioned above in the answer to 1 and in the truth tables.

3. One could write something along the lines of `2349872348723 % 7 == 0 and not(2349872348723 % 11 == 0)`. If this expression evaluated to `True` then the number would be a multiple of 7 but not of 11, if it evaluated to `False` then either it wasn't a multiple of 7 or it **was** a multiple of 11.

4. From my experiments above, I believe that the `not` operator has the highest priority, it will be evaluated first. Then the `and` operator has next highest precedence. Finally, the `or` operator has the lowest precedence of the three. It will be evaluated after the other two.

5. I would implement `xor` using the usual `and`, `or`, and `not` operators as follows. Consider two booleans `bool_one` and `bool_two`. The expression `bool_one xor bool_two` is equivalent to `(bool_one or bool_two) and not (bool_one and bool_two)`. I.e. the expression will return `True` if one of the two input booleans is `True`, but not both of them.

# Declaring variables

A central feature of programming is the declaration of variables.  When you declare a variable, you are *storing* data in the computer's *memory* and you are assigning a *name* to that data.  Both storage and name-assignment are carried out with the *single* equality symbol =.

In [1]:
e = 2.71828

With this command, the float 2.71828 is stored somewhere inside your computer, and Python can access this stored number by the name "e" thereafter.  So if you want to compute "e squared", a single command will do.

In [2]:
e * e

7.3890461584

In [3]:
type(e)

float

You can use just about any name you want for a variable, but your name *must* start with a letter, *must* not contain spaces, and your name *must* not be an existing Python word.  Characters in a variable name can include letters (uppercase and lowercase) and numbers and underscores `_`.  

So `e` is a valid name for a variable, but `type` is a bad name.  It is very tempting for beginners to use very short abbreviation-style names for variables (like `dx` or `vbn`).  But resist that temptation and use more descriptive names for variables, like `difference_x` or `very_big_number`.  This will make your code readable by you and others!

There are different style conventions for variable names.  We use lowercase names, with underscores separating words,  roughly following [Google's style conventions](https://google.github.io/styleguide/pyguide.html#Python_Style_Rules) for Python code.

In [4]:
my_number = 17

In [5]:
my_number < 23

True

After you declare a variable, its value remains the same until it is changed.  You can change the value of a variable with a simple assignment.  After the above lines, the value of my_number is 17.

In [6]:
my_number = 3.14

This command reassigns the value of my_number to 3.14.  Note that it changes the type too!  It effectively overrides the previous value and replaces it with the new value.

Often it is useful to change the value of a variable *incrementally* or *recursively*.  Python, like many programming languages, allows one to assign variables in a self-referential way.  What do you think the value of S will be after the following four lines?

In [7]:
S = 0
S = S + 1
S = S + 2
S = S + 3
print(S)

6


The first line `S = 0` is the initial declaration:  the value 0 is stored in memory, and the name S is assigned to this value.

The next line `S = S + 1` looks like nonsense, as an algebraic sentence.  But reading = as **assignment** rather than **equality**, you should read the line `S = S + 1` as assigning the *value* `S + 1` to the *name* `S`.  When Python interprets `S = S + 1`, it carries out the following steps.

1.  Compute the value of the right side, `S+1`.  (The value is 1, since `S` was assigned the value 0 in the previous line.)
2.  Assign this value to the left side, `S`.  (Now `S` has the value 1.)

Well, this is a slight lie.  Python probably does something more efficient, when given the command `S = S + 1`, since such operations are hard-wired in the computer and the Python interpreter is smart enough to take the most efficient route.  But at this level, it is most useful to think of a self-referential assignment of the form `X = expression(X)` as a two step process as above.

1.  Compute the value of `expression(X)`.
2.  Assign this value to `X`.

Now consider the following three commands.

In [8]:
my_number = 17
new_number = my_number + 1
my_number = 3.14

What are the values of the variables my_number and new_number, after the execution of these three lines?

To access these values, you can use the *print* function.

In [9]:
print(my_number)
print(new_number)

3.14
18


Python is an *interpreted* language, which carries out commands line-by-line from top to bottom.  So consider the three lines

``` python
my_number = 17
new_number = my_number + 1
my_number = 3.14
```

Line 1 sets the value of my_number to 17.  Line 2 sets the value of new_number to 18.  Line 3 sets the value of my_number to 3.14.  But Line 3 does *not* change the value of new_number at all.

(This will become confusing and complicated later, as we study mutable and immutable types.)

### Exercises

1.  What is the difference between `=` and `==` in the Python language?

2.  If the variable `x` has value `3`, and you then evaluate the Python command `x = x * x`, what will be the value of `x` after evaluation?

3.  Imagine you have two variables `a` and `b`, and you want to switch their values.  How could you do this in Python?

4.  Kepler's third law states that for a planet in circular orbit around an object of mass $M$, one has $4 \pi^2 r^3 = G M t^2$.  We can use this to estimate the mass of the sun, from other astronomically observable quantities.  Look up $G$ (the gravitational constant, estimated by experiment on Earth) and $r$ (the distance from the Earth to the sun, in meters).  Compute $t$, the number of seconds it takes for the Earth to go around the sun (365.25 days).  Finally use all of this to estimate $M$, the mass of the sun.  Your solution should use 5-10 lines of Python code.

### Solutions

(Use this space to work on the exercises.  Place more code and text cells above as needed.)

1. The `=` operator acts as an assignment operator. It will attempt to assign to the variable on the left hand side the value of the variable or expression on the right hand side. The `==` operator is a bit different. This is a boolean comparison operator which tests whether the expressions on each side of the operator are equal.

2. The value of `x` will be `9` after the evaluation of the expression. This is because `x` starts with value `3` and the line `x = x*x` computes the value `x*x = 3*3 = 9` and stores it in the variable `x`.

3. You could create a third variable `c` to temporarily store the value of one of `a` or `b`. If you try to set `a` to `b` without a temporary holding spot for the original value of `a`, it will simply be overwritten and lost. Instead, you can assign `c` to hold the value of `a`, then set `a` to the value of `b`, and finally set `b` to the value of `c`, which was originally `a`.

4. As shown in the code cell below, we compute the mass of the sun to be about $1.989\times 10^{30}$ kilograms. We use 4 significant figures since the value of G is not known to high accuracy much beyond this.

In [16]:
# Solution to Exercise 4: Computing the mass of the sun.
# Gravitational constant G
G = 6.674e-11 # Measured in m^3*kg^-1*s^-2 (rounded to 4 sig figs)
# Distance from the Earth to the Sun
r = 1.49600000e11 # Measured in meters (rounded to 4 sig figs)
# Compute t, the number of seconds it takes for the Earth to go around the sun (365.25 days)
t = 365.25*24*60*60 # Days times Hours in a day times Minutes in an hour times Seconds in a minute
# Approximate value of pi
pi = 3.14159
# Compute the mass of the sun (in kg)
M = (4*(pi**2)*(r**3)) / (G*(t**2))
print("We compute the mass of the sun to be approximately {:.3e} kilograms using 4 significant figures.".format(M))

We compute the mass of the sun to be approximately 1.989e+30 kilograms using 4 significant figures.


# Lists and ranges

Python stands out for the central role played by *lists*.  A *list* is what it sounds like -- a list of data.  Data within a list can be of any type.  Multiple types are possible within the same list!  The basic syntax for a list is to use brackets to enclose the list items and commas to separate the list items.

In [17]:
type([1,2,3])

list

In [18]:
type(['Hello',17])

list

There is another type called a *tuple* that we will use less often.  Tuples use parentheses for enclosure instead of brackets.

In [19]:
type((1,2,3))

tuple

There's another list-like type in Python 3, called the `range` type.  Ranges are kind of like lists, but instead of plunking every item into a slot of memory, ranges just have to remember three integers:  their *start*, their *stop*, and their *step*.    

The `range` command creates a range with a given start, stop, and step.  If you only input one number, the range will ***start at zero*** and use ***steps of one*** and will stop ***just before*** the given stop-number.

One can create a list from a range (plunking every term in the range into a slot of memory), by using the `list` command.  Here are a few examples.

In [20]:
type(range(10)) # Ranges are their own type, in Python 3.x.  Not in Python 2.x!

range

In [21]:
list(range(10)) # Let's see what's in the range.  Note it starts at zero!  Where does it stop?

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

A more complicated two-input form of the range command produces a range of integers **starting at** a given number, and **terminating before** another given number.

In [22]:
list(range(3,10))

[3, 4, 5, 6, 7, 8, 9]

In [23]:
list(range(-4,5))

[-4, -3, -2, -1, 0, 1, 2, 3, 4]

This is a common source of difficulty for Python beginners.  While the first parameter (-4) is the starting point of the list, the list ends just before the second parameter (5).  This takes some getting used to, but experienced Python programmers grow to like this convention.

The *length* of a list can be accessed by the len command.

In [24]:
len([2,4,6])

3

In [25]:
len(range(10))  # The len command can deal with lists and ranges.  No need to convert.

10

In [26]:
len(range(10,100)) # Can you figure out the length, before evaluating?

90

The final variant of the range command (for now) is the *three-parameter* command of the form `range(a,b,s)`.  This produces a list like `range(a,b)`, but with a "step size" of `s`.  In other words, it produces a list of integers, beginning at `a`, increasing by `s` from one entry to the next, and going up to (but not including) `b`.  It is best to experiment a bit to get the feel for it!

In [27]:
list(range(1,10,2))

[1, 3, 5, 7, 9]

In [28]:
list(range(11,30,2))

[11, 13, 15, 17, 19, 21, 23, 25, 27, 29]

In [29]:
list(range(-4,5,3))

[-4, -1, 2]

In [30]:
list(range(10,100,17))

[10, 27, 44, 61, 78, 95]

This can be used for descending ranges too, and observe that the final number b in range(a,b,s) is not included.

In [31]:
list(range(10,0,-1))

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

How many multiples of 7 are between 10 and 100?  We can find out pretty quickly with the range command and the len command (to count).

In [32]:
list(range(10,100,7))  # What list will this create?  It won't answer the question...

[10, 17, 24, 31, 38, 45, 52, 59, 66, 73, 80, 87, 94]

In [33]:
list(range(14,100,7))  # Starting at 14 gives the multiples of 7.

[14, 21, 28, 35, 42, 49, 56, 63, 70, 77, 84, 91, 98]

In [34]:
len(range(14,100,7))  # Gives the length of the list, and answers the question!

13

### Exercises

1.  If `a` and `b` are integers, what is the length of `range(a,b)`?  Express your answer as a formula involving `a` and `b`.

2.  Use a list and range command to produce the list `[1,2,3,4,5,6,7,8,9,10]`.

3.  Create the list `[1,2,3,4,5,1,2,3,4,5,1,2,3,4,5,1,2,3,4,5,1,2,3,4,5]` with a single list and range command and another operation.

4.  How many multiples of 3 are there between 300 and 3000 (including 300 and 3000)?

In [285]:
#  Use this space to work on the exercises.
# Exercise 1
print(len(range(5, 12)))

# Exercise 2
print(list(range(1, 11)))

# Exercise 3
print(list(range(1, 6))*5)

# Exercise 4
print(len(range(300, 3003, 3)))

7
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
[1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4, 5]
901


### Solutions
1. The length of `range(a,b)` is simply `b-a` since we include the number `a` in the range, but not the number `b`.
2. The command `list(range(1, 11))` gives us the desired list beginning at `1`, ending at `10`, and counting by ones.
3. The command `list(range(1, 6))*5` gives us the desired list consisting of 5 repeating copies of the elements 1 through 5.
4. There are `901` multiples of 3 between 300 and 3000 including both.

# Iterating over a range

Computers are excellent at repetitive reliable tasks.  If we wish to perform a similar computation, many times over, a computer a great tool.  Here we look at a common and simple way to carry out a repetetive computation:  the "for loop".  The "for loop" *iterates* through items in a list or range, carrying out some action for each item.  Two examples will illustrate.

In [42]:
for n in [1,2,3,4,5]:
    print(n*n)

1
4
9
16
25


In [43]:
for s in ['I','Am','Python']:
    print(s + "!")

I!
Am!
Python!


The first loop, **unraveled**, carries out the following sequence of commands.

In [44]:
n = 1
print(n*n)
n = 2
print(n*n)
n = 3
print(n*n)
n = 4
print(n*n)
n = 5
print(n*n)

1
4
9
16
25


But the "for loop" is more efficient *and* more readable to programmers.  Indeed, it saves the repetition of writing the same command `print(n*n)` over and over again.  It also makes transparent, from the beginning, the range of values that `n` is assigned to.  

When you read and write "for loops", you should consider how they look unravelled -- that is how Python will carry out the loop.  And when you find yourself faced with a repetetive task, you might consider whether it may be wrapped up in a for loop.  Most of the time, you should try to write code that *never repeats* the same task.  If there is repetition, wrap it in a loop.

Try to unravel the loop below, and predict the result, before evaluating the code.

In [45]:
P = 1
for n in range(1,6):
    P = P * n
print(P)

120


This might have been difficult!  So what if you want to trace through the loop, as it goes?  Sometimes, especially when debugging, it's useful to inspect every step of the loop to see what Python is doing.  We can inspect the loop above, by inserting a print command within the *scope* of the loop.

In [46]:
P = 1
for n in range(1,6):
    P = P * n
    print("n is",n,"and P is",P)
print(P)

n is 1 and P is 1
n is 2 and P is 2
n is 3 and P is 6
n is 4 and P is 24
n is 5 and P is 120
120


Here we have used the *print* command with strings and numbers together.  In Python 3.x, you can print multiple things on the same line by separating them by commas.  The "things" can be strings (enclosed by single or double-quotes) and numbers (int, float, etc.).

In [47]:
print("My favorite number is",17)

My favorite number is 17


If we unravel the loop above, the linear sequence of commands interpreted by Python is the following.

In [48]:
P = 1
n = 1
P = P * n
print("n is",n,"and P is",P)
n = 2
P = P * n
print("n is",n,"and P is",P)
n = 3
P = P * n
print("n is",n,"and P is",P)
n = 4
P = P * n
print("n is",n,"and P is",P)
n = 5
P = P * n
print("n is",n,"and P is",P)
print (P)

n is 1 and P is 1
n is 2 and P is 2
n is 3 and P is 6
n is 4 and P is 24
n is 5 and P is 120
120


Let's analyze the loop syntax in more detail.  
```python
P = 1
for n in range(1,6):
    P = P * n  # this command is in the scope of the loop.
    print("n is",n,"and P is",P)  # this command is in the scope of the loop too!
print(P)
```
The "for" command ends with a colon `:`, and the **next two** lines are indented.  The colon and indentation are indicators of **scope**.  The *scope* of the for loop begins after the colon, and includes all indented lines.  The *scope* of the for loop is what is repeated in every step of the loop (in addition to the reassignment of `n`).  

In [49]:
P = 1
for n in range(1,6):
    P = P * n  # this command is in the scope of the loop.
    print("n is",n,"and P is",P)  # this command is in the scope of the loop too!
print(P)

n is 1 and P is 1
n is 2 and P is 2
n is 3 and P is 6
n is 4 and P is 24
n is 5 and P is 120
120


If we change the indentation, it changes the scope of the for loop.  Predict what the following loop will do, by unraveling, before evaluating it.

In [50]:
P = 1
for n in range(1,6):
    P = P * n
print("n is",n,"and P is",P)
print(P)

n is 5 and P is 120
120


Scopes can be nested by nesting indentation.  What do you think the following loop will do?  Can you unravel it?

In [51]:
for x in [1,2,3]:
    for y in ['a', 'b']:
        print(x,y)

1 a
1 b
2 a
2 b
3 a
3 b


How might you create a nested loop which prints `1 a` then `2 a` then `3 a` then `1 b` then `2 b` then `3 b`?  Try it below.

In [52]:
# Insert your loop here.
for x in ['a', 'b']:
  for y in [1, 2, 3]:
    print(y, x)

1 a
2 a
3 a
1 b
2 b
3 b


Among popular programming languages, Python is particular about indentation.  Other languages indicate scope with open/close braces, for example, and indentation is just a matter of style.  By requiring indentation to indicate scope, Python effectively removes the need for open/close braces, and enforces a readable style.

We have now encountered data types, operations, variables, and loops.  Taken together, these are powerful tools for computation!  Now complete the following exercises for more practice.

## Exercises

1.  Describe how Python interprets division with remainder when the divisor and/or dividend is negative.
2.  What is the remainder when $2^{90}$ is divided by $91$?
3.  How many multiples of 13 are there between 1 and 1000?
4.  How many *odd* multiples of 13 are there between 1 and 1000?
5.  What is the sum of the numbers from 1 to 1000?
6.  What is the sum of the squares, from $1 \cdot 1$ to $1000 \cdot 1000$?
7.  Euler proved that
$$\frac{1}{1^4} + \frac{1}{2^4} + \frac{1}{3^4} + \cdots = \frac{\pi^4}{C},$$
for some positive integer $C$.  Use Python to guess what $C$ is.

In [79]:
# Insert your solutions here.
# Exercise 1
print("Exercise 1")
print(divmod(-100, 6)) # Positive divisor means the multiple of the divisor will be smaller than the dividend (-102=6*-17 in this case)
                       # and the remainder will be positive between 0 and divisor-1 (2 in this case)
print(divmod(-100, -6)) # Negative divisor means the multiple of the divisor will be larger than the dividend (-96=-6*16 in this case)
                        # and the remainder will be negative between 0 and divisor+1 (-4 in this case)
print()

# Exercise 2
print("Exercise 2")
# Remainder of 2**90 when divided by 91
print((2**90) % 91)
print()

# Exercise 3
print("Exercise 3")
# Number of multiples of 13 between 1 and 1000
print(len(range(13, 1000, 13)))
print()

# Exercise 4
print("Exercise 4")
# Number of odd multiples of 13 between 1 and 1000 (should be half the total number)
multiples_of_13 = list(range(13, 1000, 13))
num = 0
for item in multiples_of_13:
  if item % 2 == 0:
    num += 1
print(num)
print()

# Exercise 5
print("Exercise 5")
# What is the sum of numbers from 1 to 1000 (should be 1000*1001/2=500500 by sum formula for sequences)
sum = 0
for value in range(1, 1001, 1):
  sum += value
print(sum)
print()

# Exercise 6
print("Exercise 6")
# What is the sum of squares from 1*1 to 1000*1000
sum = 0
for value in range(1, 1001, 1):
  sum += value*value
print(sum)
print()

# Exercise 7
print("Exercise 7")
# Estimate C
# First, approximate the sequence sum using the first 1000 integers
sequence_approx = 0
for value in range(1, 1001, 1):
  sequence_approx += 1 / (value**4)
print("The sum of the infinite sequence is approximately",sequence_approx)
# Now, divide pi**4 by this approximation to approximate C
pi = 3.14159
C = pi**4 / sequence_approx
print("C is calculated to be",C, "and approximated to {:.0f}.".format(C))
print()

Exercise 1
(-17, 2)
(16, -4)

Exercise 2
64

Exercise 3
76

Exercise 4
38

Exercise 5
500500

Exercise 6
333833500

Exercise 7
The sum of the infinite sequence is approximately 1.082323233378306
C is calculated to be 89.99969594907023 and approximated to 90.



## Solutions
1. Division with remainder occurs between the dividend `A` and the divisor `B`. One gets the expression $$A = q\cdot B+r$$ where `q` is the quotient (an integer) and `r` is the remainder, an integer in the range `0` to `B-1`.

  When the divisor is positive, Python finds the *largest* multiple of the divisor `B` which is smaller than or equal to the dividend `A`. The integer that, when multiplied by `B`, results in that multiple is the quotient `q` and the remainder `r` is the positive integer between 0 and `B-1` which must be added to $q\cdot B$ in order to get `A`. This is true regardless of the sign of the dividend.

  When the divisor is negative, Python finds the *smallest* multiple of the divisor `B` which is larger than or equal to the dividend `A`. The integer that, when multiplied by `B`, results in that multiple is the quotient `q` and the remainder `r` is the *negative* integer between 0 and `B+1` which must be added to $q\cdot B$ in order to get `A`. This is true regardless of the sign of the dividend.

2. The remainder of $2^{90}$ when divided by $91$ is $64$.
3. There are $76$ multiples of $13$ between $1$ and $1000$.
4. There are $38$ *odd* multiples of $13$ between $1$ and $1000$.
5. The sum of the numbers from $1$ to $1000$ is $500500$.
6. The sum of the squares from $1\cdot 1$ to $1000 \cdot 1000$ is $333833500$.
7. Using the first 1000 terms in the sequence to approximate the sum, we find  $C$ to be $89.9997$. Thus, I would guess that $C$ is the positive integer $90$.

# Explorations

Now that you have learned the basics of computation in Python and loops, we can start exploring some interesting mathematics!  We are going to look at approximation here -- some ancient questions made easier with programming.

## Exploration 1:  Approximating square roots.

We have seen how Python can do basic arithmetic -- addition, subtraction, multiplication, and division.  But what about other functions, like the square root?  In fact, Python offers a few functions for the square root, but that's not the point.  How can we compute the square root using only basic arithmetic?  

Why might we care?

1.  We might want to know the square root of a number with more precision than the Python function offers.
2.  We might want to understand how the square root is computed... under the hood.
3.  Understanding approximations of square roots and other functions is important, because we might want to approximate other functions in the future (that aren't pre-programmed for us).

Here is a method for approximating the square root of a number $X$.

1.  Begin with a guess $g$.
2.  Observe that $g \cdot (X / g) = X$.  Therefore, among the two numbers $g$ and $(X/g)$, one will be less than or equal to the square root of X, and the other will be greater than or equal to the square root.
3.  Take the average of $g$ and $(X/g)$.  This will be closer to the square root than $g$ or $X/g$ (unless your guess is exactly right!) 
4.  Use this average as a new guess... and go back to the beginning.

Now implement this in Python to approximate the square root of 2.  Use a loop, so that you can go through the approximation process 10 times or 100 times or however many you wish.  Explore the effect of different starting guesses.  Would a change in the averaging function improve the approximation?  How quickly does this converge?  How does this change if you try square roots of different positive numbers?

Write your code (Python) and findings (in Markdown cells) in a readable form.  Answer the questions in complete sentences.



In [259]:
#  Start your explorations here!
def approximate_root(X, N):
  #g = int(X / 2)
  g = 1
  print("Starting guess:", g)
  for i in range(N):
    g = (g + (X/g))/2
    print(g)
  return g

In [282]:
approx = approximate_root(64, 15)
print("Final result:", approx)

Starting guess: 1
32.5
17.234615384615385
10.474036101145005
8.292191785986859
8.005147977880979
8.000001655289593
8.00000000000017
8.0
8.0
8.0
8.0
8.0
8.0
8.0
8.0
Final result: 8.0


Describe your findings here.
## Answers
Testing a square root approximation with starting guess 1. When we run this to calculate $\sqrt2$, we see that it converges to an answer in 5 iterations. If we use a starting guess of 5, it takes 7 iterations. A starting guess of 10 takes 8 iteration to converge. Finally, a starting guess of 100 takes 11 iterations. It looks like the number of iterations will increase slowly as the starting guess gets farther and farther away from the correct answer, but the number of iterations to convergence isn't extremely affected by the initial guess. If we wanted to make sure our initial guess was at least not ludicrously wrong, we could start at the input number divided by 2. This wouldn't be accurate by any means, but in many cases would be more accurate than always starting at 1 or 100 or 1000, for example.

One could potentially change the averaging function, but I can't think of one that would work better. In this case, the normal average works very well because it constantly chooses the guess to be directly in between the previous guess and $X/g$. This way, there's no chance of it missing and ending up with the wrong value, or blowing up to infinity. It will get closer and closer to the square root until the computer runs out of precision.

As stated before, with a starting guess of 1, this method takes 5 iterations to converge to the square root of 2. If we run it on different positive numbers, it can take more or fewer iterations depending on our initial guess and the number itself. For example, with a starting guess of 1, it takes 8 iterations to converge to the square root of 64. This is because 1 is farther off from the correct value than it was in the square root of 2 calculation. We could of course change our initial guess to be closer to the right answer, say 8 in the trivial case, and get a faster convergence. The number of iterations to converge really just depends on the difference between the initial guess and the correct answer.

## Exploration 2:  Approximating e and pi.

Now we approximate two of the most important constants in mathematics:  e and pi.  There are multiple approaches, but e is pretty easy with the series expansion of e^x.  First, approximate e by evaluating the Taylor series expansion of e^x at x=1.  I.e., remember from calculus that
$$e^x = 1 + x + \frac{1}{2} x^2 + \frac{1}{3!} x^3 + \frac{1}{4!} x^4 + \cdots.$$
How many terms are necessary before the float stabilizes?  Use a loop, with a running product for the factorials and running sums for the series.

In [266]:
#  Approximate e here.
factorial = 1
sum = 0
N = 22
for i in range(N):
  sum += 1 / factorial
  factorial *= i + 1
  print("{:>2}:    Sum: {}".format(i+1, sum))
print("Result:", sum)

 1:    Sum: 1.0
 2:    Sum: 2.0
 3:    Sum: 2.5
 4:    Sum: 2.6666666666666665
 5:    Sum: 2.708333333333333
 6:    Sum: 2.7166666666666663
 7:    Sum: 2.7180555555555554
 8:    Sum: 2.7182539682539684
 9:    Sum: 2.71827876984127
10:    Sum: 2.7182815255731922
11:    Sum: 2.7182818011463845
12:    Sum: 2.718281826198493
13:    Sum: 2.7182818282861687
14:    Sum: 2.7182818284467594
15:    Sum: 2.71828182845823
16:    Sum: 2.718281828458995
17:    Sum: 2.718281828459043
18:    Sum: 2.7182818284590455
19:    Sum: 2.7182818284590455
20:    Sum: 2.7182818284590455
21:    Sum: 2.7182818284590455
22:    Sum: 2.7182818284590455
Result: 2.7182818284590455


**Answers**

Here it seems that the loop converges to a value of $e$ after 18 iterations. If one needed $e$ to some higher precision, it would take more iterations, but we seem to have reached the highest precision we can just using python floats.

Next we will approximate pi, which is much more interesting.  We can try a few approaches.  For a series-approach (like e), we need a series that converges to pi.  A simple example is the arctangent atan(x).  Recall (precalculus!) that $atan(1) = \pi/4$.  Moreover, the derivative of $atan(x)$ is $1 / (1+x^2)$.  

1.  Figure out the Taylor series of $1 / (1+x^2)$ near $x=0$.  Note that this is a geometric series! 

2.  Figure out the Taylor series of $atan(x)$ near $x=0$ by taking the antiderivative, term by term, of the above.

3.  Try to estimate $\pi$ with this series, using many terms of the series.

**Answers**
1. The Taylor series of $1/(1+x^2)$ near $x=0$ can be represented by the geometric series
$$1-x^2+x^4-x^6+-\cdots=\sum_{n=0}^{\infty}(-x^2)^n$$

2. Taking the integral of the series above term by term gives us 
$$x-\frac{x^3}{3}+\frac{x^5}{5}-\frac{x^7}{7}+-\cdots=\sum_{n=0}^{\infty}(-1)^n\frac{x^{2n+1}}{2n+1}$$
This is the series expansion of $\arctan(x)$ near $x=0$ since $\int\frac{1}{1+x^2}\,dx=\arctan(x)$.

3. To estimate $\pi$, I will use the fact that $\arctan(1)=\pi/4$. Writing the arctangent as a series, this means that
$$\arctan(1)=\sum_{n=0}^{\infty}(-1)^n\frac{1^{2n+1}}{2n+1}=\sum_{n=0}^{\infty}(-1)^n\frac{1}{2n+1}=\frac{\pi}{4}$$
So we have the expression 
$$\pi = 4\cdot\sum_{n=0}^{\infty}(-1)^n\frac{1}{2n+1}$$
I've found that it takes an incredible 10 million terms in this series and over 3 seconds of computation to approximate $\pi$ to just 6 decimal places! This is not very efficient.

In [271]:
#  Approximate pi here!
# We have an approx for pi given by 4*(1-1/3+1/5-1/7+...)
sum = 0 
N = 10000000 # 10 Million iterations!
alternating_factor = 1
for n in range(N):
  sum += (1 / (2*n + 1))*alternating_factor
  alternating_factor *= -1
sum *= 4
print(sum)

3.1415925535897915


Now we'll accelerate things a bit.  There's a famous formula of Machin (1706) who computed the first hundred digits of $\pi$.  We'll use his identity:

$\pi/4 = 4 \cdot atan(1/5) - atan(1 / 239)$.  

This isn't obvious, but there's a tedious proof using sum/difference identities in trig.

Try using this formula now to approximate $\pi$, using your Taylor series for $atan(x)$.  It should require fewer terms.

In [272]:
#  Approximate pi more quickly here!
# Here I've written a function to use N terms of the arctangent series to approximate the value of atan(x)
def atan(x, N):
  sum = 0
  alternating_factor = 1
  for i in range(N):
    sum += (x**(2*i+1) / (2*i + 1))*alternating_factor
    alternating_factor *= -1
  return sum

In [277]:
approx = 4*(4*atan(1/5, 11) - atan(1/239, 11))
print(approx)

3.141592653589794


**Notes**

Amazingly, using Machin's formula allows us to compute $\pi$ to 15 places with just 11 terms in the approximation of each arctangent value! That is nearly a factor of 1 million fewer terms in each series when compared to the raw computation above, and it gives a much more accurate result.

Now let's compare this to **Archimedes' method**.  Archimedes approximated pi by looking at the perimeters $p(n)$ and $P(n)$ of ($2^n$)-gons inscribed in and circumscribed around a unit circle.  So $p(2)$ is the perimeter of a square inscribed in the unit circle.  $P(2)$ is the perimeter of a square circumscribed around a unit circle.

Archimedes proved the following (not in the formulaic language of algebra):  For all $n \geq 2$,

(P-formula)  $P(n+1) = \frac{2  p(n) P(n)}{p(n) + P(n)}$.

(p-formula)  $p(n+1) = \sqrt{ p(n)  P(n+1) }$.

1.  Compute $p(2)$ and $P(2)$.

2.  Use these formulas to compute $p(10)$ and $P(10)$.  Use this to get a good approximation for $\pi$!

We could use our previous sqrt function if you want, we'll take a fancier high-precision approach.  "mpmath" is a Python package for high-precision calculation.  It should be accessible through Google Colab.  You can read the full documentation at http://mpmath.org/doc/current/

First we load the package and print its status.

In [1]:
from mpmath import *
print(mp)

Mpmath settings:
  mp.prec = 53                [default: 53]
  mp.dps = 15                 [default: 15]
  mp.trap_complex = False     [default: False]


The number mp.dps is (roughly) the number of decimal digits that mpmath will keep track of in its computations.  mp.prec is the binary precision, a bit more than 3 times the decimal precision.  We can change this to whatever we want.

In [2]:
mp.dps = 50 # Let's try 50 digits precision to start.
print(mp)

Mpmath settings:
  mp.prec = 169               [default: 53]
  mp.dps = 50                 [default: 15]
  mp.trap_complex = False     [default: False]


mpmath has a nice function for square roots.  Compare this to your approximation from before!

In [3]:
sqrt(2)  # mpf(...) stands for an mp-float.  

mpf('1.4142135623730950488016887242096980785696718753769468')

In [4]:
type(sqrt(2)) # mpmath stores numbers in its own types!

mpmath.ctx_mp_python.mpf

In [5]:
4*mp.atan(1) # mpmath has the arctan built in.  This should be pretty close to pi!

mpf('3.1415926535897932384626433832795028841971693993751068')

Now try Archimedes' approximation of pi.  Use the mpmath sqrt function along the way.  How many iterations do you need to get pi correct to 100 digits?  Compare this to the arctan-series (not the mpmath atan function) via Machin's formula.

##Solutions
1. We calculate $p(2)$ to be $4\sqrt2$ and $P(2)$ to be $8$. This is because $p(2)$ calculates the perimeter of a square inscribed within the unit circle and $P(2)$ calculates the perimeter of a square circumscribed around the unit circle. These values are fairly easy to compute with simple geometry and the fact that the radius of the unit circle is 1.
2. As seen below, using the formulas above and iterating 8 times, we find that 
$$p(10) \approx 6.283175\\
P(10) \approx 6.283205$$
Where both values are rounded to 6 decimal places for clarity. The full values are printed from the code cell below.

  To approximate $\pi$ from this, we notice that both of these values are approximating the circumference of the unit circle. The more sides on the n-gon's, the closer the approximation. Also, the inscribed perimeter should be a slight underestimate, and the circumscribed perimeter should be a slight overestimate. To approximate the perimeter of the circle, we will take the average of these two values. Since the perimeter of the circle is $2\pi$, we then divide by an additional factor of 2. In summary, we say
$$\pi \approx \frac{p(10)+P(10)}{4}$$
Doing this gives us an approximation of $\pi$ accurate to 5 decimal places. See below for the code cell.

3. To approximate $\pi$ to 100 decimal places, we need to iterate the perimeter values 166 times. This means that we are using $p(168)$ and $P(168)$. Thus, we have to approximate the circumference of the circle with n-gons that have $2^{168}$ sides in order to calculate $\pi$ this accurately! Returning to the approximation of $\pi$ by Machin's formula and reworking my arctangent calculation to use the mpmath package, I found that just 71 terms in the arctangent approximation of $\arctan(\frac15)$ and $\arctan(\frac{1}{239})$ give us $\pi$ to 100 decimal places! That makes it a remarkably efficient way to approximate $\pi$.

Archimedes approximation

In [10]:
#  Explore and experiment.
# Here I calculate perimeters of n-gons
#(P-formula)  $P(n+1) = \frac{2  p(n) P(n)}{p(n) + P(n)}$.

#(p-formula)  $p(n+1) = \sqrt{ p(n)  P(n+1) }$.
# Initialize the perimeters of the inscribed and circumscribed squares
little_p = 4*mp.sqrt(2)
big_p = 8
# Iterate 8 times to go from p(2) and P(2) to p(10) and P(10)
for n in range(8):
  next_big_p = (2*little_p*big_p) / (little_p + big_p)
  little_p = mp.sqrt(little_p * next_big_p)
  big_p = next_big_p

print("p(10) = {}\nP(10) = {}".format(little_p, big_p))

p(10) = 6.2831754505543194012577085254038374787985617171497
P(10) = 6.2832050205136178935273793169853200235845469162052


In [11]:
# Approximate Pi from p(10) and P(10), seems to be accurate to 5 decimal places
pi_approx = (little_p + big_p) / 4
print(pi_approx)

3.1415951177669843236962719605972893755957771583387


In [215]:
# Set 105 points of precision, we want to approximate pi to 100 decimal places, so it's helpful to have a few extra to tell when we've gone too far
mp.dps=105
print(mp)

Mpmath settings:
  mp.prec = 352               [default: 53]
  mp.dps = 105                [default: 15]
  mp.trap_complex = False     [default: False]


In [194]:
answer_pi = 4*mp.atan(1) # Pi accurate to 105 decimal points

little_p = 4*mp.sqrt(2)
big_p = 8
# It takes about 166 iterations to approximate pi to 100 decimal places
for n in range(166):
  next_big_p = (2*little_p*big_p) / (little_p + big_p)
  little_p = mp.sqrt(little_p * next_big_p)
  big_p = next_big_p

print("Pi is approximately {}\nThis is off by {}".format((little_p + big_p) / 4, ((little_p + big_p) / 4)-answer_pi))
#print("Accurately, pi is approximately", answer_pi)

Pi is approximately 3.14159265358979323846264338327950288419716939937510582097494459230781640628620899862803482534211706800061
This is off by 1.84569546739082268799500783475163986795649726832747993957128494528461140100739038700627861836864250896757e-101


Machin's Formula

In [200]:
#  Approximating pi with Machin's formula again, this time using the mpmath package to get 100 points of accuracy
def atan_mp(x, N):
  sum = mp.mpf(0)
  alternating_factor = mp.mpf(1)
  for i in range(N):
    term = (mp.power(x, 2*i+1) / (2*i + 1))*alternating_factor
    #print(term)
    sum = sum + term
    alternating_factor *= -1
  #print(sum)
  return sum

In [227]:
# We approximate pi accurately to 100 decimal points with just 71 terms in the arctan approximations!
pi_approx = 4*(4*atan_mp(mp.mpf('1/5'), 71) - atan_mp(mp.mpf(1)/mp.mpf(239), 71))
print(pi_approx)
print("Pi is approximately {}\nThis is off by {}".format(pi_approx, pi_approx-answer_pi))

3.14159265358979323846264338327950288419716939937510582097494459230781640628620899862803482534211706799415
Pi is approximately 3.14159265358979323846264338327950288419716939937510582097494459230781640628620899862803482534211706799415
This is off by 1.20021873168286069623771185814671022761660485944699015606951786371334242128771905183250647345523664322483e-101
